src/radical/pilot/utils/prof_utils.py

```diff
@@ -1236,5 +1236,160 @@ def _get_unit_consumption(session, unit, udurations=None):
     return ret
 
 
+# ------------------------------------------------------------------------------
+#
+def get_resource_transitions(pilot, task_metrics=None, pilot_metrics=None):
+
+    if not task_metrics:
+        if pilot.cfg['task_launch_method'] == 'PRTE':
+            task_metrics = UNIT_DURATIONS_PRTE
+        else:
+            task_metrics = UNIT_DURATIONS_DEFAULT
+
+    if not pilot_metrics:
+        pilot_metrics = PILOT_DURATIONS
+
+
+    # we try to find points in time where resource usage moved from purpose A to
+    # purpose B.  For example, consider this metric:
+    #
+    #   'exec_queue'  : [{ru.EVENT: 'schedule_ok'            },
+    #                    {ru.STATE: s.AGENT_EXECUTING        }],
+    #   'exec_prep'   : [{ru.STATE: s.AGENT_EXECUTING        },
+    #                    {ru.EVENT: 'exec_start'             }],
+    #   'exec_rp'     : [{ru.EVENT: 'exec_start'             },
+    #                    {ru.EVENT: 'cu_start'               }],
+    #
+    # then we convert this into the following structure:
+    #
+    # [
+    #   # event                          from         to
+    #   [{ru.EVENT: 'schedule_ok'    },  None,        'exec_queue'],
+    #   [{ru.STATE: s.AGENT_EXECUTING}, 'exec_queue', 'exec_prep'],
+    #   [{ru.EVENT: 'exec_start'     }, 'exec_prep',  'exec_rp'],
+    #   [{ru.EVENT: 'cu_start'       }, 'exec_rp',    'None]
+    # ]
+    #
+    # which will use like this:
+    #
+    #   - go through all events for a task
+    #   - if event is in list above
+    #     - reduce resources in `from` metric
+    #     - increase resource in `to` metric
+    #
+    # If `from` or `to` are None, then the resources are taken from / given to
+    # the pilot's idle resources.  We thus rename the 'None' to 'idle'.
+
+    # dig though metric,  find all pairs of matching start/stop events
+    task_transitions = list()
+    for metric,spec in task_metrics['consume'].items():
+
+        m = None
+        for m in task_transitions:
+            if m[0] == spec[0]:
+                break
+            m = None
+        if m:
+            assert(m[2] is None)
```

Comment:
ugly :)

Reply:
Yes.  We might consider completely switching from the metric based dictionary format to the transition focused list format at some point - the reverse transition is a bit simpler.  But I added some comments for now.